File: leetcode/c++/94_binary_tree_inorder_traversal.cpp
// https://leetcode.cn/problems/binary-tree-inorder-traversal

#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode() : val(0), left(nullptr), right(nullptr) {}
      TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
      TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    vector<int> res;
    void inorder(TreeNode* root) {
        if (root == nullptr)
            return;
        inorder(root->left);
        res.push_back(root->val);
        inorder(root->right);
    }
public:
    vector<int> inorderTraversal(TreeNode* root) {
        inorder(root);
        return res;
    }
};

// O(n)，O(n)

File: leetcode/golang/94_binary_tree_inorder_traversal.go
type TreeNode struct {
         Val int
         Left *TreeNode
         Right *TreeNode
}

func inorderTraversal(root *TreeNode) []int {
    var res []int
    var inorder func(root *TreeNode) 
    inorder = func(root* TreeNode) {
        if root == nil {
            return
        }
        inorder(root.Left)
        res = append(res, root.Val)
        inorder(root.Right)
    }
    inorder(root)
    return res 
}
   